### PR TITLE
TUNIC: Additional Combat Logic Option

### PR DESCRIPTION
**Pull request by ScipioWright**
## What is this fixing or adding?
In TUNIC Rando, stat upgrades are not currently considered in logic. The idea is that you'll be receiving filler throughout the run, and you are meant to basically deal with what you have to try to beat whatever is in logic. Trying to get through Quarry? Use bombs! Get creative with Orb! etc.

This was not the most popular decision for some, especially among those who really like full-clearing their sphere 1 immediately in an async without waiting to receive any items.

This PR adds a new Combat Logic option which makes stat upgrades, some money, and a few other items progression, and requires a flexible stat requirement to get through areas, to checks, etc., with the intent of matching the vanilla difficulty of the game.

~~This PR partially contains the previous Logic Options PR: https://github.com/ArchipelagoMW/Archipelago/pull/3544
Why? Because I am _not_ going to deal with how messy of a merge it would be to do these separately.~~ - merged

## How was this tested?
Some playtesting, lots of test generations, etc.

## If this makes graphical changes, please attach screenshots.
N/A

### CODE REVIEW

File: worlds/tunic/rules.py [hunk 1]
```diff
@@ -58,18 +58,27 @@ def has_sword(state: CollectionState, player: int) -> bool:
     return state.has("Sword", player) or state.has("Sword Upgrade", player, 2)
 
 
-def has_ice_grapple_logic(long_range: bool, state: CollectionState, world: "TunicWorld") -> bool:
-    player = world.player
-    if not world.options.logic_rules:
+def laurels_zip(state: CollectionState, world: "TunicWorld") -> bool:
+    if not world.options.laurels_zips:
```
Comment (Exempt-Medic) [inline worlds/tunic/rules.py:62]: Could do this instead
```py
    return world.options.laurels_zips and state.has(laurels, world.player)
```

File: worlds/tunic/options.py [hunk 2]
```diff
@@ -187,6 +166,98 @@ class TunicPlandoConnections(PlandoConnections):
...
+    option_easy = 1
+    option_medium = 2
+    option_hard = 3
+    default = 0
+
+
+class LadderStorageWithoutItems(Toggle):
+    """
+    If disabled, you logically require Stick, Sword, or Magic Orb to Ladder Storage.
+    If enabled, you will be expected to Ladder Storage without progression items.
```
Comment (Exempt-Medic) [inline worlds/tunic/options.py:234]: Probably should say "use"?
```suggestion
    If disabled, you logically require Stick, Sword, or Magic Orb to use Ladder Storage.
    If enabled, you will be expected to use Ladder Storage without progression items.
```

File: worlds/tunic/er_rules.py [hunk 3]
```diff
@@ -1537,3 +1647,129 @@ def set_er_location_rules(world: "TunicWorld") -> None:
...
+                     lambda state, p=player, c=combat_req_area, d=dagger, la=laurel:
+                     has_combat_reqs(c, state, p)
+                     or (state.has(ice_dagger, player) if d else False)
+                     or (state.has(laurels, player) if la else False))
+        else:
+            add_rule(multiworld.get_location(loc_name, player),
+                     lambda state, p=player, c=combat_req_area, d=dagger, la=laurel:
+                     has_combat_reqs(c, state, p)
+                     or (state.has(ice_dagger, player) if d else True)
+                     or (state.has(laurels, player) if la else False))
```
Comment (Exempt-Medic) [inline worlds/tunic/er_rules.py:1667]: There's no loops here, so this is fine (same for the next one). You could also simplify some of these:
```suggestion
                     # someome tell me if you actually need to do the p=player and c=combat_req_area, lambdas scary
                     lambda state:
                     has_combat_reqs(c, state, p)
                     or (state.has(ice_dagger, player) and d)
                     or (state.has(laurels, player) and la))
        else:
            add_rule(multiworld.get_location(loc_name, player),
                     lambda state:
                     has_combat_reqs(c, state, p)
                     or (state.has(ice_dagger, player) or d)
                     or (state.has(laurels, player) and la))
```

**Review by Exempt-Medic — CHANGES_REQUESTED** [review]
Two small comments, I looked through stuff that wasn't `combat_logic.py` that'll be for another day. Haven't done generations, but the rest of the code made sense and matched my understanding of the game.

File: worlds/tunic/options.py (hunk 2 above)
Comment (ScipioWright) [reply worlds/tunic/options.py:234]: I'm using Ladder Storage as a verb here. I'll run it by some others in the Tunic community and see what they think

**Review by Exempt-Medic — APPROVED** [review]
The changes LGTM, various comments in DMs were addressed. Merged into `main` and did ten thousand players worth of generations, using various numbers of players and did not hit any errors. Something I guess core should be aware of is that this uses `ItemClassification` with multiple at once, which like, this _should_ be fine, but it's worth mentioning / keeping in mind.

**Review by roseasromeo — APPROVED** [review]
Combat logic makes sense based on my experience playing the game. Indirect conditions for entrances are accounted for (when accessing an area may require diving into a nearby entrance to deaggro enemies). Ran a few test generations with both boss combat logic and full combat logic and spoilers look reasonable.

I have also previously tested combat logic in runs with grass rando, and the results worked well.

File: worlds/tunic/combat_logic.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+    unchecked = 0
+    failed = 1
+    succeeded = 2
+
+
+def has_combat_reqs(area_name: str, state: CollectionState, player: int) -> bool:
+    # we're caching whether you've met the combat reqs before if the state didn't change first
+    # if the combat state is stale, mark each area's combat state as stale
+    if state.tunic_need_to_reset_combat_from_collect[player]:
+        state.tunic_need_to_reset_combat_from_collect[player] = 0
```
[the remark this answers is not in the document]
Comment (ScipioWright) [reply worlds/tunic/combat_logic.py:66]: Probably just leftover from the many iterations this had -- I'll change those o:

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**